msos/WaitChainTraversal.cs

```diff
@@ -0,0 +1,216 @@
+
+            result.SetInfo(info);
+
+            return result;
+        }
+
+        private void HandleWctRequestError()
+        {
+            var lastErrorCode = Marshal.GetLastWin32Error();
+
+            if (lastErrorCode == (uint)SYSTEM_ERROR_CODES.ERROR_IO_PENDING)
+            {
+                //TODO: Follow this doc: https://msdn.microsoft.com/en-us/library/windows/desktop/ms681421(v=vs.85).aspx
+            }
+            else
+            {
+                //TODO : Ifdentify code error and responce accordingly
+            }
+        }
+
+        private void CheckCount(ref int Count)
+        {
+            // Check if the wait chain is too big for the array we passed in.
+            if (Count > WCT_MAX_NODE_COUNT)
+            {
+                //Found additional nodes 
+                Count = WCT_MAX_NODE_COUNT;
+            }
+        }
+    }
+
+    class ThreadWCTInfo
+    {
+        public ThreadWCTInfo(bool isDeadLock, uint threadId)
+        {
+            IsDeadLocked = isDeadLock;
+            ThreadId = threadId;
+        }
+
+        /// <summary>
+        /// Specifies whether the Wait Chain is Cyclic - Deadlock
+        /// </summary>
+        public bool IsDeadLocked { get; private set; }
+        /// <summary>
+        /// OS Id of the thread
+        /// </summary>
+        public uint ThreadId { get; private set; }
+
+        List<WaitChainInfoObject> _blockingObjects;
+
+        public List<WaitChainInfoObject> WctBlockingObjects
+        {
+            get
+            {
+                if (_blockingObjects == null)
+                {
+                    _blockingObjects = new List<WaitChainInfoObject>();
+                }
+                return _blockingObjects;
+            }
+        }
+
+        internal void SetInfo(WAITCHAIN_NODE_INFO[] info)
+        {
+            if (info != null)
+            {
+                foreach (var item in info)
+                {
+                    var block = new WaitChainInfoObject(item);
+                    WctBlockingObjects.Add(block);
+                }
+            }
+        }
+    }
+
+    class WaitChainInfoObject
+    {
+        public WaitChainInfoObject(WAITCHAIN_NODE_INFO item)
+        {
+            ObjectStatus = item.ObjectStatus;
+            ObjectType = item.ObjectType;
+
+
+            //LockObject stuct data
+            TimeOut = item.Union.LockObject.Timeout;
+            AlertTable = item.Union.LockObject.Alertable;
+
+            if (item.ObjectType == WCT_OBJECT_TYPE.WctThreadType)
+            { //Use the ThreadObject part of the union
+                this.ThreadId = item.Union.ThreadObject.ThreadId;
+                this.ProcessId = item.Union.ThreadObject.ProcessId;
+                this.ContextSwitches = item.Union.ThreadObject.ContextSwitches;
+                this.WaitTime = item.Union.ThreadObject.WaitTime;
+
+            }
+            else
+            {//Use the LockObject  part of the union
+                unsafe
```

Comment:
I don't actually understand why, but when I change it to IntPtr the process is stuck on reading WAITCHAIN_NODE_INFO structure cause it contains a union of _WAITCHAIN_NODE_INFO_UNION structure...